pkg/logs/internal/decoder/auto_multiline_detection/pattern_table.go

```diff
@@ -0,0 +1,146 @@
+// The patternt table is always sorted by the frequency of the patterns. When the table
+// becomes full, the least recently updated pattern is evicted.
+type PatternTable struct {
+	table                           []*row
+	index                           int64
+	maxTableSize                    int
+	matchThreshold                  float64
+	dontAggregateUnmatchedTopFormat bool
+}
+
+// NewPatternTable returns a new PatternTable heuristic.
+func NewPatternTable(maxTableSize int, matchThreshold float64, dontAggregateUnmatchedTopFormat bool) *PatternTable {
+	return &PatternTable{
+		table:                           make([]*row, 0, maxTableSize),
+		index:                           0,
+		maxTableSize:                    maxTableSize,
+		matchThreshold:                  matchThreshold,
+		dontAggregateUnmatchedTopFormat: dontAggregateUnmatchedTopFormat,
+	}
+}
+
+func (p *PatternTable) insert(tokens []tokens.Token, label Label) int {
+	p.index++
+	foundIdx := -1
+	for i, r := range p.table {
+		if isMatch(r.tokens, tokens, p.matchThreshold) {
+			r.count++
+			r.label = label
+			r.lastIndex = p.index
+			foundIdx = i
+			break
+		}
+	}
+
+	if foundIdx == 0 {
+		return foundIdx
+	}
+
+	if foundIdx > 0 {
+		return p.siftUp(foundIdx)
+	}
+
+	// If the table is full, make room for a new entry
+	if len(p.table) >= p.maxTableSize {
+		p.evictLRU()
+	}
+
+	p.table = append(p.table, &row{
+		tokens:    tokens,
+		label:     label,
+		count:     1,
+		lastIndex: p.index,
+	})
+	return len(p.table) - 1
+
+}
+
+// siftUp moves the row at the given index up the table until it is in the correct position.
+func (p *PatternTable) siftUp(idx int) int {
+	for idx != 0 && p.table[idx].count > p.table[idx-1].count {
+		p.table[idx], p.table[idx-1] = p.table[idx-1], p.table[idx]
+		idx--
+	}
+	return idx
+}
+
+// evictLRU removes the least recently updated row from the table.
+func (p *PatternTable) evictLRU() {
+	mini := 0
+	minIndex := p.index
+	for i, r := range p.table {
+		if r.lastIndex < minIndex {
+			minIndex = r.lastIndex
+			mini = i
+		}
+	}
+	p.table = append(p.table[:mini], p.table[mini+1:]...)
+}
+
+// DumpTable returns a slice of DiagnosticRow structs that represent the current state of the table.
+func (p *PatternTable) DumpTable() []DiagnosticRow {
+	debug := make([]DiagnosticRow, 0, len(p.table))
+	for _, r := range p.table {
+		debug = append(debug, DiagnosticRow{
+			TokenString: tokensToString(r.tokens),
+			LabelString: labelToString(r.label),
+			Count:       r.count,
+			LastIndex:   r.lastIndex})
+	}
+	return debug
+}
+
+// Process adds a pattern to the table and updates its label based on it's frequency.
+// This implements the Herustic interface - so we should stop processing if the label was changed
+// due to pattern detection.
+func (p *PatternTable) Process(context *messageContext) bool {
+
+	if context.tokens == nil {
+		log.Error("Tokens are required to process patterns")
+		return true
```

Comment:
Does it makes sense to return true, when there was no process being made?

Reply:
I should probably rename this function to make it more clear what it's return type means. [The comment here should help](https://github.com/DataDog/datadog-agent/blob/038ebf16431127643594c8e28a7e5ca68ecb72c8/pkg/logs/internal/decoder/auto_multiline_detection/labeler.go#L31)
`true` means keep processing (move on to the next heuristic)
`false` means stop processing (we have determined the best label for this log)

While not having tokens is an error case here (that shouldn't ever happen at runtime) it doesn't mean that future heuristics in the pipeline are invalid.

Reply:
WDYT of naming it `ProcessAndContinue`? I think that makes it more clear what `return true` or `false` means contextually?

Reply:
`ProcessAndContinue ` sounds great 😄 

Thanks for the explanation